api/explorer.go

```diff
@@ -1,62 +1,334 @@
-func (srv *Server) explorerGetHashHandler(w http.ResponseWriter, req *http.Request) {
-	// Extract the hash from the request
-	var data []byte
-	_, err := fmt.Sscanf(req.FormValue("hash"), "%x", &data)
-	if err != nil {
-		writeError(w, "Malformed or no hash provided: "+err.Error(), http.StatusBadRequest)
+	// Try the hash as an unlock hash.
+	txids := srv.explorer.UnlockHash(types.UnlockHash(hash))
+	if len(txids) != 0 {
+		var txns []ExplorerTransaction
+		var blocks []ExplorerBlock
+		for _, txid := range txids {
+			block, height, exists := srv.explorer.Transaction(txid)
+			if !exists && build.DEBUG {
+				panic("explorer pointing to nonexistant txn")
+			}
+			if types.TransactionID(block.ID()) == txid {
+				blocks = append(blocks, ExplorerBlock{
+					ID:       block.ID(),
+					Height:   height,
+					RawBlock: block,
+				})
+			}
+			for _, t := range block.Transactions {
+				if t.ID() == txid {
+					txns = append(txns, ExplorerTransaction{
+						ID: txid,
+						Parent: ExplorerBlock{
+							ID:       block.ID(),
+							Height:   height,
+							RawBlock: block,
+						},
+						RawTransaction: t,
+					})
+				}
+			}
+		}
+		writeJSON(w, ExplorerHashGET{
+			HashType:     "unlockhash",
+			Blocks:       blocks,
+			Transactions: txns,
+		})
 		return
 	}
 
-	// returnData will be a generic interface. The json encoder
-	// should still work though
-	returnData, err := srv.exp.GetHashInfo(data)
-	if err != nil {
-		writeError(w, err.Error(), http.StatusInternalServerError)
+	// Try the hash as a siacoin output id.
+	txids = srv.explorer.SiacoinOutputID(types.SiacoinOutputID(hash))
+	if len(txids) != 0 {
+		var txns []ExplorerTransaction
+		var blocks []ExplorerBlock
+		for _, txid := range txids {
+			block, height, exists := srv.explorer.Transaction(txid)
+			if !exists && build.DEBUG {
+				panic("explorer pointing to nonexistant txn")
+			}
+			if types.TransactionID(block.ID()) == txid {
+				blocks = append(blocks, ExplorerBlock{
+					ID:       block.ID(),
+					Height:   height,
+					RawBlock: block,
+				})
+			}
+			for _, t := range block.Transactions {
+				if t.ID() == txid {
+					txns = append(txns, ExplorerTransaction{
+						ID: txid,
+						Parent: ExplorerBlock{
+							ID:       block.ID(),
+							Height:   height,
+							RawBlock: block,
+						},
+						RawTransaction: t,
+					})
+				}
+			}
+		}
+		writeJSON(w, ExplorerHashGET{
+			HashType:     "siacoinoutputid",
+			Blocks:       blocks,
+			Transactions: txns,
+		})
+		return
+	}
+
+	// Try the hash as a file contract id.
+	txids = srv.explorer.FileContractID(types.FileContractID(hash))
+	if len(txids) != 0 {
+		var txns []ExplorerTransaction
+		var blocks []ExplorerBlock
+		for _, txid := range txids {
+			block, height, exists := srv.explorer.Transaction(txid)
+			if !exists && build.DEBUG {
+				panic("explorer pointing to nonexistant txn")
+			}
+			if types.TransactionID(block.ID()) == txid {
+				blocks = append(blocks, ExplorerBlock{
```

Comment:
The constructions for the api and the explorer are different. The explorer module is designed to provide the information to other Go programs. As such, it doesn't need to worry about providing extra information such as the block id. Just call block.ID() and you can get that information yourself.

The API on the other hand provides data to programs that don't have access to functions like ID(), and so needs to include a lot of extra information that's really irrelevant when using the go interfaces directly.

Additionally, it's okay to make a bunch of function calls. It's not okay though to make a bunch of API calls, because API calls are slow, especially if you aren't making calls from the same machine (the explorer is going to be exposed over the internet in many cases, which means making API calls is super slow). The API calls should provide a lot more information to minimize the number of calls that need to be made.